src/client/render/pipeline.h

```diff
@@ -0,0 +1,391 @@
+};
+
+class DynamicTarget : public RenderTarget
+{
+public:
+	bool isConfigured() { return upstream != nullptr; }
+	void setRenderTarget(RenderTarget *value) { upstream = value; }
+	virtual void activate(PipelineContext *context) override;
+private:
+	RenderTarget *upstream { nullptr };
+};
+
+/**
+ * Base class for rendering steps in the pipeline
+ */
+class RenderStep : virtual public RenderPipelineObject
+{
+public:
+	/**
+	 * Assigns render source to this step.
+	 * 
+	 * @param source source of rendering information
+	 */
+	virtual void setRenderSource(RenderSource *source) = 0;
+
+	/**
+	 * Assigned render target to this step.
+	 * 
+	 * @param target render target to send output to.
+	 */
+	virtual void setRenderTarget(RenderTarget *target) = 0;
+
+	/**
+	 * Runs the step. This method is invoked by the pipeline.
+	 */
+	virtual void run(PipelineContext *context) = 0;
+};
+
+/**
+ * Provides default empty implementation of supporting methods in a rendering step.
+ */
+class TrivialRenderStep : public RenderStep
+{
+public:
+	virtual void setRenderSource(RenderSource *source) override {}
+	virtual void setRenderTarget(RenderTarget *target) override {}
+	virtual void reset(PipelineContext *) override {}
+};
+
+/**
+ * Dynamically changes render target of another step.
+ * 
+ * This allows re-running parts of the pipeline with different outputs
+ */
+class SetRenderTargetStep : public TrivialRenderStep
+{
+public:
+	SetRenderTargetStep(RenderStep *step, RenderTarget *target);
+	virtual void run(PipelineContext *context) override;
+private:
+	RenderStep *step;
+	RenderTarget *target;
+};
+
+/**
+ * Render Pipeline provides a flexible way to execute rendering steps in the engine.
+ * 
+ * RenderPipeline also implements @see RenderStep, allowing for nesting of the pipelines.
+ */
+class RenderPipeline : public RenderStep
+{
+public:
+	/**
+	 * Add a step to the end of the pipeline
+	 * 
+	 * @param step reference to a @see RenderStep implementation.
+	 */
+	void addStep(RenderStep *step)
+	{
+		m_pipeline.push_back(step);
+	}
+
+	/**
+	 * Capture ownership of a dynamically created @see RenderStep instance.
+	 * 
+	 * RenderPipeline will delete the instance when the pipeline is destroyed.
+	 * 
+	 * @param step reference to the instance.
+	 * @return RenderStep* value of the 'step' parameter.
+	 */
+	template<class T>
+	T *own(T *object)
+	{
+		m_objects.push_back(std::unique_ptr<RenderPipelineObject>(object));
+		return object;
+	}
+
+	RenderSource *getInput();
+	RenderTarget *getOutput();
+
+	v2f getScale() { return scale; }
+	void setScale(v2f value) { scale = value; }
+
+	virtual void reset(PipelineContext *context) override {}
+	virtual void run(PipelineContext *context) override;
+
+	virtual void setRenderSource(RenderSource *source) override;
+	virtual void setRenderTarget(RenderTarget *target) override;
+private:
+	std::vector<RenderStep *> m_pipeline;
+	std::vector< std::unique_ptr<RenderPipelineObject> > m_objects;
```

Comment:
I think you want a set here in case someone ever calls own twice

Reply:
This is a compile-time problem, and set has worse run-time performance. I can add a debug-only check if necessary.